src/main/kotlin/com/nikitanazarov/coroutinestacks/CoroutineStacksPanel.kt

```diff
@@ -1,64 +1,386 @@
+                    it.color = color
+                    it.drawRoundRect(x, y, width - 1, height - 1, arc, arc)
+                }
             }
         }
 
-        coroutineGraph = mxGraphComponent(graph)
-        add(coroutineGraph)
+        return roundedBorder
+    }
+
+
+    private fun createPanelBuilderListener() = object : DebugProcessListener {
+        override fun paused(suspendContext: SuspendContext) {
+            emptyText.component.isVisible = false
+            buildCoroutineGraph(suspendContext)
+        }
+
+        override fun resumed(suspendContext: SuspendContext?) {
+            panelContent.removeAll()
+        }
+    }
+
+    fun buildCoroutineGraph(suspendContext: SuspendContext) {
+        val suspendContextImpl = suspendContext as? SuspendContextImpl ?: run {
+            emptyText.text = CoroutineStacksBundle.message("coroutine.stacks.could.not.be.built")
+            return
+        }
+
+        val coroutineInfoCache = CoroutineDebugProbesProxy(suspendContextImpl).dumpCoroutines()
+        val coroutineInfoDataList = coroutineInfoCache.cache
+        val dispatcherToCoroutineDataList = mutableMapOf<String, MutableList<CoroutineInfoData>>()
+        for (info in coroutineInfoDataList) {
+            val dispatcher = info.descriptor.dispatcher ?: continue
+            dispatcherToCoroutineDataList.computeIfAbsent(dispatcher) { mutableListOf() }.add(info)
+        }
+
+        val dispatcherLabel = JLabel(CoroutineStacksBundle.message("select.dispatcher"))
+        val coroutineStacksWindowHeader = Box.createHorizontalBox()
+        val dispatchers = dispatcherToCoroutineDataList.keys.toTypedArray()
+        val dispatcherDropdownMenu = ComboBox(dispatchers)
+
+        dispatcherDropdownMenu.addActionListener {
+            val selectedDispatcher = dispatcherDropdownMenu.selectedItem as String
+            val coroutineDataList = dispatcherToCoroutineDataList[selectedDispatcher]
+            if (!coroutineDataList.isNullOrEmpty()) {
+                updateCoroutineStackForest(coroutineDataList, suspendContextImpl)
+            }
+        }
+
+        val comboBoxSize = Dimension(Constants.comboBoxHeight, Constants.comboBoxWidth)
+        dispatcherDropdownMenu.apply {
+            preferredSize = comboBoxSize
+            maximumSize = comboBoxSize
+            minimumSize = comboBoxSize
+        }
+        coroutineStacksWindowHeader.add(dispatcherLabel)
+        coroutineStacksWindowHeader.add(dispatcherDropdownMenu)
+
+        dispatcherToCoroutineDataList.forEach { (dispatcher, coroutineDataList) ->
+            coroutineDataList.forEach { data ->
+                if (data.descriptor.state == State.RUNNING) {
+                    dispatcherDropdownMenu.selectedItem = dispatcher
+                }
+            }
+        }
+
+        val selectedDispatcher = dispatcherDropdownMenu.selectedItem as String
```

Comment:
This comment is still pending

Reply:
Why you have marked this conversation as resolved while it is not resolved? This line contains a bug. The `javax.swing.JComboBox#getSelectedItem` function is marked as `@Nullable`, which means it can return null. So if selectedItem is null, strong casting it to String will throw an exception!